### DISCUSSION TITLE
nested triples count

### DISCUSSION THREAD

**Discussion by majidaldo** [oxigraph/oxigraph#443] [Q&A] [answered]
**Describe the bug**
not sure this counts as a 'bug' but there needs to be a way to get the length of reified triples.

**To Reproduce**
```
s = og.Store()
s.update("""
prefix :<test://>
insert data {<<<<:s :p :o>>  :p  <<:s :p :o>>  >>  :p :o.}
"""
)
print(
len(s)
)
```
gives 1. but the reified triples are 5. (uggh are repeated triples counted??).

**Comment by majidaldo** [comment]
note: i couldn't find a recursing sparql query to get the count.

**Comment by Tpt** [comment, marked as answer]
Hi! In [RDF-star](https://w3c.github.io/rdf-star/cg-spec/editors_draft.html) triples used as subject or object are called "quoted-triples" and are not directly inserted in the graph. For example having `<< :sq :pq :oq >> :p :o` in the graph does not entails that  `:sq :pq :oq` is in the graph. Ence the result for the `len` function that is coherent with what `SELECT (COUNT(*) AS ?c) WHERE { GRAPH ?g { ?s ?p ?o }}` would return.

> note: i couldn't find a recursing sparql query to get the count.

I don't think that SPARQL-star allows to recurse on nested triples in the general case, sadly.

An example of query with bounded nesting:
```sparql
SELECT (COUNT(DISTINCT ?triple) AS ?c) {
  { ?s ?p ?o BIND(TRIPLE(?s, ?p, ?o) AS ?triple) }
  UNION { ?triple ?p ?o FILTER(isTriple(?triple)) }
  UNION { ?s ?p ?triple FILTER(isTriple(?triple)) }
  ...
}
```

**Comment by majidaldo** [comment]
> Hi! In [RDF-star](https://w3c.github.io/rdf-star/cg-spec/editors_draft.html) triples used as subject or object are called "quoted-triples" and are not directly inserted in the graph. For example having `<< :sq :pq :oq >> :p :o` in the graph does not entails that `:sq :pq :oq` is in the graph. Ence the result for the `len` function that is coherent with what `SELECT (COUNT(*) AS ?c) WHERE { GRAPH ?g { ?s ?p ?o }}` would return.


would they `__hash__` differently?

**Reply by Tpt** [reply]
In python the `Triple` object is used both for quoted triples and triples inside of the graph so, yes, they get the same hash.

**Reply by majidaldo** [reply]
...but they're different  :/

is this some unresolved rdfstar issue?

**Reply by Tpt** [reply]
Sorry, I am not sure to properly understand your answer. Which "they" are you referring to?

**Reply by majidaldo** [reply]
sorry.

I was referring to the the hash of  Triple(NestedTriple1, P, O) vs Triple(NestedTriple2, P, O). And would the same difference show up on the store hash.

**Reply by Tpt** [reply]
Thank you! So if `NestedTriple1 != NestedTriple2` then `hash( Triple(NestedTriple1, P, O))` should be different from `hash(Triple(NestedTriple2, P, O))`. The hashing works recursively on nested triples.

**Reply by majidaldo** [reply]
but the db object doesn't hash as i would expect

```
 In [4]: s=Store()
In [5]: hash(s)
Out[5]: 95114994587
In [7]: s.update("insert data {<a://> <b://> <c://> } ")
In [8]: hash(s)
Out[8]: 95114994587
```

**Reply by majidaldo** [reply]
store.__hash__ = hash(frozenset(store))  ?

**Reply by Tpt** [reply]
I have not implemented `__hash__` (and `__eq__`) on `Store` because they are mutable objects. So, you get the default `__hash__` implementation from Python that is useless. It's sad the default is not to raise an exception instead.

**Reply by majidaldo** [reply]
> I have not implemented `__hash__` (and `__eq__`) on `Store` because they are mutable objects. So, you get the default `__hash__` implementation from Python that is useless. It's sad the default is not to raise an exception instead.

well i'm expecting this calculation to come from rust.

**Reply by Tpt** [reply]
> well i'm expecting this calculation to come from rust.

Yes, and they are not implemented in Rust either. The main assumption of the `Store` class is that it stores millions of triples so hashing it is not an expected operation.

**Reply by majidaldo** [reply]
> > well i'm expecting this calculation to come from rust.
> 
> Yes, and they are not implemented in Rust either. The main assumption of the `Store` class is that it stores millions of triples so hashing it is not an expected operation.

but then how is adding redundant data handled?

**Reply by Tpt** [reply]
> but then how is adding redundant data handled?

The `Store` uses BTrees internally. So, it requires an ordering operation to be defined on RDF terms/triples but does not at all requires the `Store` object to build a value-base hash.

**Comment by majidaldo** [comment]
(can be moved to discussion)f
